### ISSUE TITLE
Auto mount params as props

### ISSUE THREAD

**Issue by rottmann** [rackt/react-router#591]
How can the router params automatically set as props on each route?
I think i did not see the forest for the tree ;)

I create a Projectview with Navigation and Dashboard.
A click on the Navigation link replace the Dashboard with the Detailview of an ID.

In the Detailview i want to access the ID with `this.props.id`, but it can only be done if it is explict defined in Project RouteHandler. 
For me it is an unnecessary dependence in the component, because i have to look then in the router-definition what params are needed.

I don't want to access the router params in the Detailview directly, i don't want to think about what router or how the router is used in a component, `this.props` exists for this case.

See my comment in var Project

```
var Project = React.createClass({
    render: function() {
        return (
            <div className="project">
                <ProjectNavigation />
                <RouteHandler />
/*
This would be correct on route to project-details, but not on route to projects-dashboard:

                var id = this.getParams().id ? parseInt(this.getParams().id, 10) : 0;
                <RouteHandler id={id} />

Why must it be set explicit? It could be automatically transfered from the router.
*/
            </div>
        );
    }
});

var ProjectNavigation = React.createClass({
    /*...*/
    render: function() {
        var items = this.state.projects.map(function(item) {
            return (
                <li key={item.id}>
                    <Link to="project-details" params={{ id: item.id }}>{item.name}</Link>
                </li>
            );
        });
        return (
            <div className="project-navigation">
                <ul>{items}</ul>
            </div>
        );
    }
});

var ProjectsDashboard = React.createClass({
    render: function() {
        return (
            <div className="projects-dashboard">
                Select a project in navigation.
            </div>
        );
    }
});

var ProjectDetails = React.createClass({
    propTypes: {
        id: React.PropTypes.number.isRequired
    },
    /*...*/
    render: function() {
        return (
            <div className="project-details">
                Project-ID: {this.props.id}
            </div>
        );
    }
});

var routes = (
    <Route handler={App}>
        <Route name="project" handler={Project}>
            <Route name="projects-dashboard" path="/projects/dashboard" handler={ProjectsDashboard} />
            <Route name="project-details" path="/project/:id" handler={ProjectDetails} />
            <DefaultRoute handler={ProjectsDashboard} />
        </Route>
    </Route>
);

Router.run(routes, function(Handler, state) {
    React.render(<Handler flux={flux} />, document.getElementById('app'));
});
```

**Comment by bitmage** [comment]
Your [documentation](https://github.com/rackt/react-router/blob/master/docs/guides/path-matching.md) seems to indicate that `this.props.params` should be automatically set.  Like @rottmann I'm also encountering the behavior that it is not set, and I have to include the `mixins: [Router.State]` and call `this.getParams()` in order to access parameters.  In addition I notice that if I call `this.getParams()` inside `getDefaultProps` no params are available.  (They *are* available inside the render or getInitialState methods.)  So it's not easy to shim the functionality as a mixin.  I am testing with the latest NPM version (0.11.6).

Can we get some clarification on how this is supposed to work?

**Comment by gaearon** [comment]
@bitmage Thanks a lot for reporting.

Docs indeed seem outdated wrt `this.props.params`. Would you like to submit a PR correcting them with `State` mixin usage instead?

>In addition I notice that if I call this.getParams() inside getDefaultProps no params are available.

This is by design because React as of 0.12 (I think) calls `getDefaultProps` only once and caches it for all instances. So this can't be made to work anyway.

**Comment by gaearon** [comment]
@rottmann 

You can do `<RouteHandler {...this.getParams()} />` everywhere and have them proxied. It's a new syntax called JSX spread attributes.

**Comment by contra** [comment]
This is still slightly confusing though, how would a situation like this work?

```xml
<Route name="project-details" path="/project/:id" handler={ProjectDetails}>
    <Route name="issue-details" path="/issue/:id" handler={IssueDetails} />
</Route>
```

This URL won't work `/project/123/issue/456` (as far as I can tell)

`this.getParams()` in ProjectDetails and IssueDetails will return the same value since it isn't contextual. It would be nice to have a function that only returns relevant parameters to your context within the route. For example, `this.getParams()` in ProjectDetails would return `{id: '123'}` and `this.getParams()` in IssueDetails would return `{id: '456'}`

**Comment by gaearon** [comment]
@contra

Your configuration is a bit wrong btw.

```js
<Route name="project-details" path="/project/:id" handler={ProjectDetails}>
    <Route name="issue-details" path="/issue/:id" handler={IssueDetails} />
</Route>
```

This would mean `/issue/:id` is *the* path for child route. It starts with slash so it is considered absolute and replaces parent's path. Router doesn't mandate path nesting so you can do that, but in this case `id` is the only parameter you'll get (`ProjectDetails will also get this `id`).

```js
<Route name="project-details" path="/project/:id" handler={ProjectDetails}>
    <Route name="issue-details" path="issue/:id" handler={IssueDetails} />
</Route>
```

This exhibits the problem you're talking about: `id` is ambiguous. A warning might help here. I don't think `params` should be contextual because this wouldn't work, for example, with linking. How do you do `<Link to='issue-details' params={{ projectId: 100, issueId: 200 }}>` if those `id`s are contextual? Instead, you should name them differently:

```js
<Route name="project-details" path="/project/:projectId" handler={ProjectDetails}>
    <Route name="issue-details" path="issue/:issueId" handler={IssueDetails} />
</Route>
```

**Comment by contra** [comment]
@gaearon Edited my comment to correct that - thanks

**Comment by mjackson** [comment]
@contra It's better to not have ambiguous params, as @gaearon suggests. We should add a warning to make this more obvious.

**Comment by contra** [comment]
@mjackson Agree

**Comment by rpflorence** [comment]
I think @contra should add a warning, closing this in lieu of his pull request :P

**Comment by rpflorence** [comment]
also, I'll be reviewing the docs soon, looking for the automatic props stuff.
